### ISSUE TITLE
A Quick Question About Training SeerAttention-Decoding

### ISSUE THREAD

**Issue by zjnyly** [microsoft/SeerAttention#38]
Hi, I followed the training script to start the process, but due to limited resources (a single RTX 3090), I had to make several modifications to the configurations. As a result, I’m unsure whether the training is proceeding as expected.

I noticed that the paper mentions distillation takes approximately 10 hours using two MI300x GPUs, while the default 800 training steps in the script don't seem to cover even a single full epoch over the generated dataset in my case.

Given the limited GPU memory, I made the following adjustments:
```
training_max_length=${TRAINING_MAX_LENGTH:-8192}
gpus=${GPUS:-1}
bs=${BS:-4}
gradient_accumulation_steps=$((1))
```
Additionally, I had to filter out sequences longer than TRAINING_MAX_LENGTH to avoid out-of-memory errors:
```
def build_optimized_chunks(dataset, chunk_size):
    samples = []
    num_to_load = len(dataset) // 16
    # for fast validation
    dataset = dataset.select(range(num_to_load))
    for example in dataset:
        input_ids = torch.as_tensor(example["input_ids"])
        seq_len = input_ids.size(0)
        # samples.append((input_ids, seq_len))
        # MODIFIED HERE
        if seq_len <= 8192:
            samples.append((input_ids, seq_len))
```
I also increased the number of steps to 1600, but the overall training time remains relatively short—likely because the chunk size is limited to 8192 tokens. Here’s a sample of the training log:
```
{'loss': 0.4222, 'grad_norm': 0.12177972495555878, 'learning_rate': 1.0140897297377359e-05, 'epoch': 0.7}                                                          
{'loss': 0.3823, 'grad_norm': 0.10966914147138596, 'learning_rate': 1.0118393765081136e-05, 'epoch': 0.7}                                                          
{'loss': 0.8872, 'grad_norm': 0.20185448229312897, 'learning_rate': 1.009784676381755e-05, 'epoch': 0.7}                                                           
{'loss': 0.4152, 'grad_norm': 0.12838034331798553, 'learning_rate': 1.0079256374819762e-05, 'epoch': 0.7}                                                          
{'loss': 0.4333, 'grad_norm': 0.151651993393898, 'learning_rate': 1.0062622671585594e-05, 'epoch': 0.7}                                                            
{'loss': 0.4231, 'grad_norm': 0.15384554862976074, 'learning_rate': 1.0047945719876875e-05, 'epoch': 0.7}                                                          
{'loss': 0.4133, 'grad_norm': 0.11295422166585922, 'learning_rate': 1.0035225577719428e-05, 'epoch': 0.7}                                                          
{'loss': 0.384, 'grad_norm': 0.11437777429819107, 'learning_rate': 1.002446229540276e-05, 'epoch': 0.7}                                                            
{'loss': 0.4159, 'grad_norm': 0.148076131939888, 'learning_rate': 1.0015655915479938e-05, 'epoch': 0.7}                                                            
{'loss': 0.5441, 'grad_norm': 0.1604791283607483, 'learning_rate': 1.0008806472767206e-05, 'epoch': 0.7}                                                           
{'loss': 0.5488, 'grad_norm': 0.13536547124385834, 'learning_rate': 1.0003913994344094e-05, 'epoch': 0.7}                                                          
{'loss': 0.4684, 'grad_norm': 0.1313875913619995, 'learning_rate': 1.0000978499553155e-05, 'epoch': 0.7}                                                           
{'train_runtime': 2742.4762, 'train_samples_per_second': 0.583, 'train_steps_per_second': 0.583, 'train_loss': 0.6421144595742225, 'epoch': 0.7}                   
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1600/1600 [45:42<00:00,  1.71s/it]
```
I have two main questions:

1. At what loss range would it be reasonable to consider the training converged or sufficiently complete for distillation?

2. Do my modifications significantly harm the training process? My concern is that filtering out long sequences and using a small batch size may reduce the model's ability to handle long-context inputs.

Any insights or suggestions would be greatly appreciated!

**Comment by YizhaoGao** [comment]
> Hi, I followed the training script to start the process, but due to limited resources (a single RTX 3090), I had to make several modifications to the configurations. As a result, I’m unsure whether the training is proceeding as expected.
> 
> I noticed that the paper mentions distillation takes approximately 10 hours using two MI300x GPUs, while the default 800 training steps in the script don't seem to cover even a single full epoch over the generated dataset in my case.
> 
> Given the limited GPU memory, I made the following adjustments:
> 
> ```
> training_max_length=${TRAINING_MAX_LENGTH:-8192}
> gpus=${GPUS:-1}
> bs=${BS:-4}
> gradient_accumulation_steps=$((1))
> ```
> 
> Additionally, I had to filter out sequences longer than TRAINING_MAX_LENGTH to avoid out-of-memory errors:
> 
> ```
> def build_optimized_chunks(dataset, chunk_size):
>     samples = []
>     num_to_load = len(dataset) // 16
>     # for fast validation
>     dataset = dataset.select(range(num_to_load))
>     for example in dataset:
>         input_ids = torch.as_tensor(example["input_ids"])
>         seq_len = input_ids.size(0)
>         # samples.append((input_ids, seq_len))
>         # MODIFIED HERE
>         if seq_len <= 8192:
>             samples.append((input_ids, seq_len))
> ```
> 
> I also increased the number of steps to 1600, but the overall training time remains relatively short—likely because the chunk size is limited to 8192 tokens. Here’s a sample of the training log:
> 
> ```
> {'loss': 0.4222, 'grad_norm': 0.12177972495555878, 'learning_rate': 1.0140897297377359e-05, 'epoch': 0.7}                                                          
> {'loss': 0.3823, 'grad_norm': 0.10966914147138596, 'learning_rate': 1.0118393765081136e-05, 'epoch': 0.7}                                                          
> {'loss': 0.8872, 'grad_norm': 0.20185448229312897, 'learning_rate': 1.009784676381755e-05, 'epoch': 0.7}                                                           
> {'loss': 0.4152, 'grad_norm': 0.12838034331798553, 'learning_rate': 1.0079256374819762e-05, 'epoch': 0.7}                                                          
> {'loss': 0.4333, 'grad_norm': 0.151651993393898, 'learning_rate': 1.0062622671585594e-05, 'epoch': 0.7}                                                            
> {'loss': 0.4231, 'grad_norm': 0.15384554862976074, 'learning_rate': 1.0047945719876875e-05, 'epoch': 0.7}                                                          
> {'loss': 0.4133, 'grad_norm': 0.11295422166585922, 'learning_rate': 1.0035225577719428e-05, 'epoch': 0.7}                                                          
> {'loss': 0.384, 'grad_norm': 0.11437777429819107, 'learning_rate': 1.002446229540276e-05, 'epoch': 0.7}                                                            
> {'loss': 0.4159, 'grad_norm': 0.148076131939888, 'learning_rate': 1.0015655915479938e-05, 'epoch': 0.7}                                                            
> {'loss': 0.5441, 'grad_norm': 0.1604791283607483, 'learning_rate': 1.0008806472767206e-05, 'epoch': 0.7}                                                           
> {'loss': 0.5488, 'grad_norm': 0.13536547124385834, 'learning_rate': 1.0003913994344094e-05, 'epoch': 0.7}                                                          
> {'loss': 0.4684, 'grad_norm': 0.1313875913619995, 'learning_rate': 1.0000978499553155e-05, 'epoch': 0.7}                                                           
> {'train_runtime': 2742.4762, 'train_samples_per_second': 0.583, 'train_steps_per_second': 0.583, 'train_loss': 0.6421144595742225, 'epoch': 0.7}                   
> 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1600/1600 [45:42<00:00,  1.71s/it]
> ```
> 
> I have two main questions:
> 
> 1. At what loss range would it be reasonable to consider the training converged or sufficiently complete for distillation?
> 2. Do my modifications significantly harm the training process? My concern is that filtering out long sequences and using a small batch size may reduce the model's ability to handle long-context inputs.
> 
> Any insights or suggestions would be greatly appreciated!

Hi, thanks for the question. 
1. About number of training tokens used: Yes, you don't have to use the entire dataset. It's a rather large one. 
2. Convergence: typically you will see 12-15x loss reduction compared to loss at first step. I think your number looks ok.
3. Potential problem with training config: the only concern is about your config is the max length 8k. It would be better to have some longer sequence in the training. But the performance might still be ok because the use of rope. You can try more different config.

BTW, you might also want to update with our inference modeling, we have fixed a few bugs recently.

**Comment by zjnyly** [comment]
> > Hi, I followed the training script to start the process, but due to limited resources (a single RTX 3090), I had to make several modifications to the configurations. As a result, I’m unsure whether the training is proceeding as expected.
> > I noticed that the paper mentions distillation takes approximately 10 hours using two MI300x GPUs, while the default 800 training steps in the script don't seem to cover even a single full epoch over the generated dataset in my case.
> > Given the limited GPU memory, I made the following adjustments:
> > ```
> > training_max_length=${TRAINING_MAX_LENGTH:-8192}
> > gpus=${GPUS:-1}
> > bs=${BS:-4}
> > gradient_accumulation_steps=$((1))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Additionally, I had to filter out sequences longer than TRAINING_MAX_LENGTH to avoid out-of-memory errors:
> > ```
> > def build_optimized_chunks(dataset, chunk_size):
> >     samples = []
> >     num_to_load = len(dataset) // 16
> >     # for fast validation
> >     dataset = dataset.select(range(num_to_load))
> >     for example in dataset:
> >         input_ids = torch.as_tensor(example["input_ids"])
> >         seq_len = input_ids.size(0)
> >         # samples.append((input_ids, seq_len))
> >         # MODIFIED HERE
> >         if seq_len <= 8192:
> >             samples.append((input_ids, seq_len))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I also increased the number of steps to 1600, but the overall training time remains relatively short—likely because the chunk size is limited to 8192 tokens. Here’s a sample of the training log:
> > ```
> > {'loss': 0.4222, 'grad_norm': 0.12177972495555878, 'learning_rate': 1.0140897297377359e-05, 'epoch': 0.7}                                                          
> > {'loss': 0.3823, 'grad_norm': 0.10966914147138596, 'learning_rate': 1.0118393765081136e-05, 'epoch': 0.7}                                                          
> > {'loss': 0.8872, 'grad_norm': 0.20185448229312897, 'learning_rate': 1.009784676381755e-05, 'epoch': 0.7}                                                           
> > {'loss': 0.4152, 'grad_norm': 0.12838034331798553, 'learning_rate': 1.0079256374819762e-05, 'epoch': 0.7}                                                          
> > {'loss': 0.4333, 'grad_norm': 0.151651993393898, 'learning_rate': 1.0062622671585594e-05, 'epoch': 0.7}                                                            
> > {'loss': 0.4231, 'grad_norm': 0.15384554862976074, 'learning_rate': 1.0047945719876875e-05, 'epoch': 0.7}                                                          
> > {'loss': 0.4133, 'grad_norm': 0.11295422166585922, 'learning_rate': 1.0035225577719428e-05, 'epoch': 0.7}                                                          
> > {'loss': 0.384, 'grad_norm': 0.11437777429819107, 'learning_rate': 1.002446229540276e-05, 'epoch': 0.7}                                                            
> > {'loss': 0.4159, 'grad_norm': 0.148076131939888, 'learning_rate': 1.0015655915479938e-05, 'epoch': 0.7}                                                            
> > {'loss': 0.5441, 'grad_norm': 0.1604791283607483, 'learning_rate': 1.0008806472767206e-05, 'epoch': 0.7}                                                           
> > {'loss': 0.5488, 'grad_norm': 0.13536547124385834, 'learning_rate': 1.0003913994344094e-05, 'epoch': 0.7}                                                          
> > {'loss': 0.4684, 'grad_norm': 0.1313875913619995, 'learning_rate': 1.0000978499553155e-05, 'epoch': 0.7}                                                           
> > {'train_runtime': 2742.4762, 'train_samples_per_second': 0.583, 'train_steps_per_second': 0.583, 'train_loss': 0.6421144595742225, 'epoch': 0.7}                   
> > 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1600/1600 [45:42<00:00,  1.71s/it]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I have two main questions:
> > 
> > 1. At what loss range would it be reasonable to consider the training converged or sufficiently complete for distillation?
> > 2. Do my modifications significantly harm the training process? My concern is that filtering out long sequences and using a small batch size may reduce the model's ability to handle long-context inputs.
> > 
> > Any insights or suggestions would be greatly appreciated!
> 
> Hi, thanks for the question.
> 
> 1. About number of training tokens used: Yes, you don't have to use the entire dataset. It's a rather large one.
> 2. Convergence: typically you will see 12-15x loss reduction compared to loss at first step. I think your number looks ok.
> 3. Potential problem with training config: the only concern is about your config is the max length 8k. It would be better to have some longer sequence in the training. But the performance might still be ok because the use of rope. You can try more different config.
> 
> BTW, you might also want to update with our inference modeling, we have fixed a few bugs recently.

Thanks for your reply. Really excellent job, and I will keep following!
